src/object-store/src/util.rs

```diff
@@ -78,7 +78,55 @@ pub fn normalize_dir(v: &str) -> String {
 /// - Otherwise, it's a file path.
 pub fn join_path(parent: &str, child: &str) -> String {
     let output = format!("{parent}/{child}");
-    opendal::raw::normalize_path(&output)
+    normalize_path(&output)
+}
+
+/// Modified from the `opendal::raw::normalize_path`
+///
+/// # The different
+///
+/// It only keeps `/` ahead if the original path starts with `/`.
+///
+/// Make sure all operation are constructed by normalized path:
+///
+/// - Path endswith `/` means it's a dir path.
+/// - Otherwise, it's a file path.
+///
+/// # Normalize Rules
+///
+/// - All whitespace will be trimmed: ` abc/def ` => `abc/def`
+/// - **(Removed❗️)** ~~ All leading / will be trimmed: `///abc` => `abc`~~
```

Comment:
We don't need this diff comment. You know it because it comes from OpenDAL and we do the diff, but it's meaningless for following developers.

Also, I'm unsure if we should normalize `///abc` to ///abc` or `/abc`. Somehow our inputs should be clean (only `/abc` as input) and we need not worry about it anyway.

@v0y4g3r @evenyag do you remember what behavior we want here? It looks like quite a tribal rule and I'd like to know where/how GreptimeDB depends on this join function.

Reply:
We could replace most places that use `format!("{}/{}")` to join two object paths with this helper function. We might get `/a//b` if inputs are `/a/` and `b` as `format` always insert a `/` into the result. We can't use `std::path::Path` as we need a path for object stores so we still use `/` as delimiters on Windows.

> Somehow our inputs should be clean (only /abc as input) and we need not worry about it anyway.

We might deal with user inputs so we should sanitize the path.

e.g. The path might be retrieved from a `COPY DATABASE` SQL statement.
https://github.com/GreptimeTeam/greptimedb/pull/3164#discussion_r1461390626

Reply:
> We don't need this diff comment.

This follows the comment in `normalize_dir()`. Do we still need them? @v0y4g3r

Reply:
> I'm unsure if we should normalize ///abc to ///abc or /abc

This is the desired behavior.

> We don't need this diff comment.

We don't need to explain the difference between this function and opendal's. @dalprahcd just keep the documentation for normilze rules.